File: crates/verify/ess-conformance/tests/fixtures/stored-field-guards-adversary-runtime.mjs
import test from 'node:test';
import { run } from './dist/runtime.js';
const refuses = (row) => {
  const weight = Number(String(row.weight_kg));
  const express = row.service === 'Express';
  switch (process.env.ESS_PARCELS_MUTANT) {
    case 'weight': return express;
    case 'fields': return false;
    case 'service': return weight > 20;
    default: return express && weight > 20;
  }
};
class Parcels {
  rows = new Map();
  minted = 0;
  identity() { return { name: 'parcels', version: '1' }; }
  beginScenario() { this.rows = new Map(); }
  endScenario() {}
  configureExternalOutcome() { throw Error('not externally forced'); }
  executeCommand({ command, input }) {
    this.minted += 1;
    const result = { consistency: String(this.minted) };
    if (command === 'shipping.parcel.Create') {
      const id = `00000000-0000-4000-8000-${String(this.minted).padStart(12, '0')}`;
      this.rows.set(id, { ...input, parcel_id: id, state: 'Created' });
      result.outcome = 'created';
      result.directEvents = [{ event: 'shipping.parcel.Created', payload: { parcel_id: id } }];
    } else if (command === 'shipping.parcel.Dispatch') {
      const row = this.rows.get(input.parcel_id);
      if (!row || row.state !== 'Created') return result;
      if (refuses(row)) {
        result.outcome = 'refused-overweight';
        result.error = 'shipping.parcel.ExpressOverweight';
        return result;
      }
      row.state = 'Dispatched';
      result.outcome = 'dispatched';
      result.directEvents = [{ event: 'shipping.parcel.Dispatched', payload: {} }];
    } else throw Error(`unknown command ${command}`);
    return result;
  }
  queryView() { return { rows: [...this.rows.values()] }; }
  observeEvents() { return []; }
  redeliverEvent() { throw Error('no bindings'); }
  observeInvocations() { return []; }
}
await test('parcels', (t) => run(t, () => new Parcels()));

File: crates/verify/ess-conformance/tests/fixtures/stored-field-guards-adversary-runtime.go
package essconform

import (
	"fmt"
	"os"
	"testing"
)

type parcels struct {
	rows   map[string]map[string]Node
	minted int
}

func (*parcels) Identity() (Identity, error) { return Identity{Name: "parcels", Version: "1"}, nil }
func (p *parcels) BeginScenario(ScenarioContext) error {
	p.rows = map[string]map[string]Node{}
	return nil
}
func (*parcels) EndScenario(ScenarioContext) error { return nil }
func (*parcels) ConfigureExternalOutcome(ExternalOutcomeControl) error {
	return fmt.Errorf("not externally forced")
}

func refuses(row map[string]Node) bool {
	weight, _ := asNumber(row["weight_kg"])
	express := row["service"] == "Express"
	switch os.Getenv("ESS_PARCELS_MUTANT") {
	case "weight":
		return express
	case "fields":
		return false
	case "service":
		return weight > 20
	}
	return express && weight > 20
}

func (p *parcels) ExecuteCommand(r CommandRequest) (CommandResult, error) {
	p.minted++
	result := CommandResult{Consistency: fmt.Sprintf("%d", p.minted)}
	switch r.Command {
	case "shipping.parcel.Create":
		id := fmt.Sprintf("00000000-0000-4000-8000-%012d", p.minted)
		row := map[string]Node{}
		for k, v := range r.Input {
			row[k] = v
		}
		row["parcel_id"] = id
		row["state"] = "Created"
		p.rows[id] = row
		result.Outcome = "created"
		result.DirectEvents = []ObservedEvent{{Event: "shipping.parcel.Created", Payload: map[string]Node{"parcel_id": id}}}
	case "shipping.parcel.Dispatch":
		id, _ := r.Input["parcel_id"].(string)
		row, ok := p.rows[id]
		if !ok || row["state"] != "Created" {
			return result, nil
		}
		if refuses(row) {
			result.Outcome = "refused-overweight"
			result.Error = "shipping.parcel.ExpressOverweight"
			return result, nil
		}
		row["state"] = "Dispatched"
		result.Outcome = "dispatched"
		result.DirectEvents = []ObservedEvent{{Event: "shipping.parcel.Dispatched", Payload: map[string]Node{}}}
	default:
		return result, fmt.Errorf("unknown command %s", r.Command)
	}
	return result, nil
}

func (p *parcels) QueryView(ViewRequest) (ViewResult, error) {
	rows := []Row{}
	for _, row := range p.rows {
		rows = append(rows, row)
	}
	return ViewResult{Rows: rows}, nil
}
func (*parcels) ObserveEvents(EventObservationRequest) ([]ObservedEvent, error) { return nil, nil }
func (*parcels) RedeliverEvent(RedeliveryRequest) error                        { return ErrUnsupported }
func (*parcels) ObserveInvocations(InvocationObservationRequest) ([]Invocation, error) {
	return nil, ErrUnsupported
}
func TestParcels(t *testing.T) { Run(t, func() Target { return &parcels{} }) }
